// @matthewscharles/cm-toolbox/index.js
// /**
//  * CM tools
//  * @namespace
//  */

/**
 * classes
 */

import { CSSToggle } from './src/classes/CSSToggle.js'


console.log('running cm-toolbox');

/**
 * methods
 */

import { addAnimation } from './src/methods/addAnimation.js'
import { constrain } from './src/methods/constrain.js'
import { create } from './src/methods/create.js'
import { createFilter } from './src/methods/createFilter.js'
import { createAnimation } from './src/methods/createAnimation.js'
import { createNS } from './src/methods/createNS.js'
import { emit } from './src/methods/emit.js'
import { ftos } from './src/methods/ftos.js'
import { getTranslate } from './src/methods/transform.js'
import { hexToInt } from './src/methods/hexToInt.js'
import { interpolate } from './src/methods/interpolate.js'
import { intToHex } from './src/methods/intToHex.js'
import { map } from './src/methods/map.js'
import { mtof } from './src/methods/mtof.js'
import { mtos } from './src/methods/mtos.js'
import { populate } from './src/methods/populate.js'
import { rot } from './src/methods/rot.js'
import { set } from './src/methods/set.js'
import { setInTree } from './src/methods/setInTree.js'
import { setNS } from './src/methods/setNS.js'
import { stom } from './src/methods/stom.js'
import { stof } from './src/methods/stof.js'
import { toggleFullscreen } from './src/methods/toggleFullscreen.js'
import { transposeForPlayer } from './src/methods/transposeForPlayer.js'
import { v } from './src/methods/v.js'
import { within } from './src/methods/within.js'

import { generateSVGFilter, generateFeColorMatrix, generateFeComponentTransfer, generateAnimate } from './src/svg/svg_filter_generators.js'
import { addFiltersToDefs, filterFromDictionary, applyFilter } from './src/svg/svg_filters.js'
import { setSvgTextClass } from './src/svg/setSvgTextClass.js'
import { hexToRGB, RGBToHSL, HSLToHex, floatToHexDigit, hexDigitToFloat } from './src/methods/colours.js'

export {
    CSSToggle,

    addAnimation,
    constrain,
    create,
    createFilter,
    createAnimation,
    createNS,
    emit,
    ftos,
    getTranslate,
    hexToInt,
    interpolate,
    intToHex,
    map,
    mtof,
    mtos,
    populate,
    rot,
    set,
    setInTree,
    setNS,
    stom,
    stof,
    toggleFullscreen,
    transposeForPlayer,
    v,
    within,

    generateSVGFilter,
    generateFeColorMatrix,
    generateFeComponentTransfer,
    generateAnimate,
    addFiltersToDefs,
    filterFromDictionary,
    applyFilter,

    setSvgTextClass,

    hexToRGB,
    RGBToHSL,
    HSLToHex,
    floatToHexDigit,
    hexDigitToFloat


}

// define CM for export default
var CM = {
    CSSToggle,

    addAnimation,
    constrain,
    create,
    createFilter,
    createAnimation,
    createNS,
    emit,
    ftos,
    getTranslate,
    hexToInt,
    interpolate,
    intToHex,
    map,
    mtof,
    mtos,
    populate,
    rot,
    set,
    setInTree,
    setNS,
    stom,
    stof,
    toggleFullscreen,
    transposeForPlayer,
    v,
    within,

    generateSVGFilter,
    generateFeColorMatrix,
    generateFeComponentTransfer,
    generateAnimate,
    addFiltersToDefs,
    filterFromDictionary,
    applyFilter,

    setSvgTextClass,

    hexToRGB,
    RGBToHSL,
    HSLToHex,
    floatToHexDigit,
    hexDigitToFloat
}

if (typeof module !== 'undefined' && module.exports) {
    console.log('cm-toolbox running in node...success!')
    module.exports = CM;
} else if(typeof globalThis !== 'undefined') {
    globalThis.CM = CM;
} else if(typeof window !== 'undefined') {
    window.CM = CM;
}

export default CM
